src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/rewriting/RuleApplicationBaseImpl.java

```diff
@@ -0,0 +1,64 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.physical.BinaryPhysicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.NaryPhysicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.UnaryPhysicalOp;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalPlanWithBinaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalPlanWithNaryRootImpl;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.PhysicalPlanWithUnaryRootImpl;
+
+public abstract class RuleApplicationBaseImpl implements RuleApplication{
+    protected final PhysicalPlan[] subPlans;
+    protected int index;
+
+    public RuleApplicationBaseImpl( final PhysicalPlan[] subPlans ){
+        this.subPlans = subPlans;
+    }
+
+    @Override
+    public PhysicalPlan getPlan() {
+        return subPlans[0];
+    }
+
+    @Override
+    public PhysicalPlan getResultingPlan() {
+        index = 0;
+        return constructPlan( subPlans[0] );
+    }
+
+    protected PhysicalPlan constructPlan( final PhysicalPlan plan ) {
+        final PhysicalPlan[] newSubPlans = getRewrittenSubPlans(plan);
```

Comment:
This doesn't look right. You don't need to (try to) rewrite all subplans recursively. Instead, begin with the last plan in `subPlans` because that's the subplan to which the actual rewriting is meant to be applied. After that, you can walk your way up in the overall plan (by going over `subPlans` in the reverse order---last to first) to recreate the parent subplans all the way back to the overall root.